File: padroesComportamentais/padrao-de-projeto-observer/php/tests/unit/UsuarioTest.php
<?php

namespace unit;

use Observer\Php\Loja;
use Observer\Php\Produto;
use Observer\Php\Usuario;
use PHPUnit\Framework\TestCase;

class UsuarioTest extends TestCase
{
    public function testDeveNotificarUsuario()
    {
        $usuario = new Usuario();
        $loja = (new Loja())->attach($usuario);
        $produto = (new Produto())->setNome("camisa")->setPreco(10.0)->setPromocao(10);
        $loja->setProdutosEmPromocao($produto)->notify();
        $this->assertEquals("nome do produto: camisa | preco: 9", $usuario->getNotificacaoDesconto());
    }

    public function testDeveNotificarUsuarios()
    {
        $usuario = new Usuario();
        $usuario2 = new Usuario();
        $loja = (new Loja())->attach($usuario)->attach($usuario2);
        $produto = (new Produto())->setNome("camisa")->setPreco(10.0)->setPromocao(10);
        $loja->setProdutosEmPromocao($produto)->notify();
        $this->assertEquals("nome do produto: camisa | preco: 9", $usuario->getNotificacaoDesconto());
        $this->assertEquals("nome do produto: camisa | preco: 9", $usuario2->getNotificacaoDesconto());
    }

    public function testNaoDeveNotificarUsuario()
    {
        $usuario = new Usuario();
        $loja = new Loja();
        $produto = (new Produto())->setNome("camisa")->setPreco(10.0)->setPromocao(10);
        $loja->setProdutosEmPromocao($produto)->notify();
        $this->assertEquals("", $usuario->getNotificacaoDesconto());
    }

    public function testDeveNotificarSomenteUsuarioPrimeiraLoja()
    {
        $usuario = new Usuario();
        $usuario2 = new Usuario();
        $loja = (new Loja())->attach($usuario);
        $loja2 = (new Loja())->attach($usuario2);
        $produto = (new Produto())->setNome("camisa")->setPreco(10.0)->setPromocao(10);
        $loja->setProdutosEmPromocao($produto)->notify();
        $this->assertEquals("nome do produto: camisa | preco: 9", $usuario->getNotificacaoDesconto());
        $this->assertEquals("", $usuario2->getNotificacaoDesconto());
    }
}

File: padroesComportamentais/padrao-de-projeto-observer/java/src/test/java/UsuarioTest.java
import observer.Loja;
import observer.Produto;
import observer.Usuario;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class UsuarioTest {

    @Test
    public void deveNotificarUsuario() {
        Loja loja = new Loja();
        Usuario usuario = new Usuario();
        usuario.observaPromocao(loja);
        loja.notificaPromocao(new Produto().setNome("camisa").setPreco(10f).setPromocao(10f));
        assertEquals("nome do produto: camisa | preco: 9.0", usuario.getMensagemPromocao());
    }

    @Test
    public void deveNotificarUsuarios() {
        Loja loja = new Loja();
        Usuario usuario = new Usuario();
        Usuario usuario2 = new Usuario();
        usuario.observaPromocao(loja);
        usuario2.observaPromocao(loja);
        loja.notificaPromocao(new Produto().setNome("camisa").setPreco(10f).setPromocao(10f));
        assertEquals("nome do produto: camisa | preco: 9.0", usuario.getMensagemPromocao());
        assertEquals("nome do produto: camisa | preco: 9.0", usuario2.getMensagemPromocao());
    }

    @Test
    public void naoDeveNotificarUsuario() {
        Loja loja = new Loja();
        Usuario usuario = new Usuario();

        loja.notificaPromocao(new Produto().setNome("camisa").setPreco(10f).setPromocao(10f));
        assertEquals("", usuario.getMensagemPromocao());
    }

    @Test
    public void deveNotificarSomenteUsuarioPrimeiraLoja() {
        Loja loja = new Loja();
        Loja loja2 = new Loja();
        Usuario usuario = new Usuario();
        Usuario usuario2 = new Usuario();
        usuario.observaPromocao(loja);
        usuario2.observaPromocao(loja2);
        loja.notificaPromocao(new Produto().setNome("camisa").setPreco(10f).setPromocao(10f));
        assertEquals("nome do produto: camisa | preco: 9.0", usuario.getMensagemPromocao());
        assertEquals("", usuario2.getMensagemPromocao());
    }
}
